### PR TITLE
[AF-543] Integrate Perspective Editor on Library Structure

### PR DESCRIPTION
**Pull request by ederign**

### CODE REVIEW

**Comment by ederign** [comment]
depends on https://github.com/AppFormer/uberfire/pull/765

**Comment by ederign** [comment]
@Rikkola two things:
- My new resource type is never indexed. Should I do something more else?
- At least on my machine, new projects are never indexed on the UI (but looks like they are indexed):
screenshot 2017-07-07 19 18 59

**Comment by manstis** [comment]
Indexing is shafted at the moment. Changes in AppFormer relating to ExecutorService have broken all async processing. A fix is in progress. porcelli has details.

**Comment by manstis** [comment]
What do you mean by your new resource type is not indexed? Async jobs are currently broken on master  so Library is likely not to show items (it uses Lucene indexes to get the list). However if the scope of "impact analysis" it is recommended a new Indexer implementation is written to store "useful" information about (new) asset types.

**Comment by Rikkola** [comment]
@ederign Yes we have a default indexer that is used if you don't make your own. That should pick it up.
I would try making a DRL or POJO file for example, just to check if what @manstis described is the issue.

**Comment by ederign** [comment]
jenkins retest this

**Comment by ederign** [comment]
That is why I cannot see my new resource type. I think the default one is fine. We are green now! :)

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/pom.xml]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/model/PerspectiveExtensionContent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/model/PerspectiveExtensionContent.java]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/model/PerspectiveExtensionContent.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    private LayoutTemplate content;
+    private Overview overview;
+
+    public PerspectiveExtensionContent() {
+    }
+
+    public PerspectiveExtensionContent(final LayoutTemplate content,
+                                       final Overview overview
+    ) {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/model/PerspectiveExtensionContent.java]: Want to move the closing bracket?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/type/PerspectiveExtensionResourceTypeDefinition.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.uberfire.backend.vfs.Path;
+import org.uberfire.workbench.type.ResourceTypeDefinition;
+
+@ApplicationScoped
+public class PerspectiveExtensionResourceTypeDefinition
+        implements ResourceTypeDefinition {
+
+    @PostConstruct
+    public void bla() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/java/org/drools/workbench/screens/appformer/extensions/type/PerspectiveExtensionResourceTypeDefinition.java]: Probably not needed?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/resources/ErraiApp.properties [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/resources/ErraiApp.properties]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/resources/org/drools/workbench/screens/appformer/extensions/AppformerExtensionsEditorAPI.gwt.xml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!--
+  ~ Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-api/src/main/resources/org/drools/workbench/screens/appformer/extensions/AppformerExtensionsEditorAPI.gwt.xml]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/pom.xml [hunk 7]
```diff
@@ -0,0 +1,115 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/pom.xml]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+    private SafeSessionInfo safeSessionInfo;
+
+    @Inject
+    private LayoutServices layoutServices;
+
+    public PerspectiveExtensionServiceImpl() {
+    }
+
+    @Inject
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java]: Best to be consistent with injection; either field or constructor (and constructor is going to make your life easier for Unit Testing.. which, given the field instances are ```private``` I suspect there's no tests for this class.. but I'll find out soon enough).

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        }
+    }
+
+    @Override
+    public PerspectiveExtensionContent loadContent(final Path path) {
+        return super.loadContent(path);
+    }
+
+    @Override
+    protected PerspectiveExtensionContent constructContent(Path path,
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java]: It's nice to be consistent with (or without) ```final``` modifiers on method parameters.. but it's an opinion so ignore me!

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+                                                           Overview overview) {
+
+        String content = load(path);
+        LayoutTemplate layoutTemplate = layoutServices.convertLayoutFromString(content);
+
+        resourceOpenedEvent.fire(new ResourceOpenedEvent(path,
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImpl.java]: We should probably convince ourselves these events are still used. They were when we have an "Inbox" feature; however that was removed in 7.0.0.Final. I am less certain however whether "Social Activities" still use these events. If they're not used by "Social" please could you create a GUVNOR JIRA saying we can remove use of "Resource Opened Events" (and assign to me if you like).

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/resources/ErraiApp.properties [hunk 11]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/main/resources/ErraiApp.properties]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PerspectiveExtensionServiceImplTest {
+
+    private IOService ioService = mock(IOService.class);
+    private KieProjectService projectService = mock(KieProjectService.class);
+
+    @InjectMocks
+    private PerspectiveExtensionServiceImpl perspectiveExtensionService = new PerspectiveExtensionServiceImpl();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImplTest.java]: I take it back.... there is a test :-)

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+    private String filename = "dora.perspective.extension";
+    private Path path2 = PathFactory.newPath("contextpath", "file:///contextpath");
+
+    @Test
+    public void testCreateAlreadyExisting() throws URISyntaxException {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/java/org/drools/workbench/screens/appformer/extensions/backend/PerspectiveExtensionServiceImplTest.java]: OK, one test.. one minimal test.. I guess the service doesn't really do much anyway.. but if you're writing one minimal test; is it worth writing some more!?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/resources/META-INF/services/org.uberfire.java.nio.file.spi.FileSystemProvider [hunk 14]
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright 2014 JBoss, by Red Hat, Inc
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-backend/src/test/resources/META-INF/services/org.uberfire.java.nio.file.spi.FileSystemProvider]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/pom.xml [hunk 15]
```diff
@@ -0,0 +1,122 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/pom.xml]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/AppFormerExtensionsEditorEntryPoint.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.drools.workbench.screens.appformer.extensions.client;
+
+import org.jboss.errai.ioc.client.api.AfterInitialization;
+import org.jboss.errai.ioc.client.api.EntryPoint;
+
+@EntryPoint
+public class AppFormerExtensionsEditorEntryPoint {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/AppFormerExtensionsEditorEntryPoint.java]: Any need for this class? 

You're not declaring an ```errai-ui``` ```@Bundle``` or anything (although perhaps the client is not i18n'ed... let me look)...

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenter.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    private PerspectiveExtensionResourceType perspectiveExtensionResourceType;
+
+    private Caller<PerspectiveExtensionService> perspectiveExtensionService;
+
+    private Caller<LayoutServices> layoutServices;
+
+    private LayoutEditorPresenter layoutEditorPresenter;
+
+    @Inject
+    public PerspectiveExtensionEditorPresenter(final PerspectiveExtensionEditorView view,
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenter.java]: Please keep some level of consistency of ```final``` usage (I don't mind which!)

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorView.css [hunk 18]
```diff
@@ -0,0 +1,10 @@
+.pe-container {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorView.css]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.html [hunk 19]
```diff
@@ -0,0 +1,4 @@
+<div class="pe-container">
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.html]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        extends KieEditorViewImpl
+        implements PerspectiveExtensionEditorView {
+
+    @Inject
+    @DataField("pe-layout")
+    Div layoutContainer;
+    private PerspectiveExtensionEditorPresenter presenter;
+
+    @Override
+    public void onResize() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java]: Needed?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public void init(final PerspectiveExtensionEditorPresenter presenter) {
+        this.presenter = presenter;
+    }
+
+    @Override
+    protected void onLoad() {
+        super.onLoad();
+        Scheduler.get().scheduleDeferred(() -> onResize());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorViewImpl.java]: Not really needed is it (given the implementation of ```onResize()```)?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/handler/NewPerspectiveExtensionHandler.java [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+        this.perspectiveExtensionService = perspectiveExtensionService;
+        this.perspectiveExtensionResourceTypeDefinition = perspectiveExtensionResourceTypeDefinition;
+        this.busyIndicatorView = busyIndicatorView;
+    }
+
+    @Override
+    public String getDescription() {
+
+        return AppformerExtensionsEditorConstants.INSTANCE.newPerspectiveExtension();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/handler/NewPerspectiveExtensionHandler.java]: Should you not be using Errai's ```TranslationService```?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/resources/i18n/AppformerExtensionsEditorConstants.java [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+package org.drools.workbench.screens.appformer.extensions.client.resources.i18n;
+
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.i18n.client.Messages;
+
+public interface AppformerExtensionsEditorConstants
+        extends
+        Messages {
+
+    AppformerExtensionsEditorConstants INSTANCE = GWT.create(AppformerExtensionsEditorConstants.class);
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/resources/i18n/AppformerExtensionsEditorConstants.java]: Should you not be using Errai's ```TranslationService```?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/resources/images/AppFormerExtensionsEditorImageResources.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/java/org/drools/workbench/screens/appformer/extensions/client/resources/images/AppFormerExtensionsEditorImageResources.java]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/ErraiApp.properties [hunk 26]
```diff
@@ -0,0 +1,31 @@
+#
+# Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/ErraiApp.properties]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/META-INF/beans.xml [hunk 27]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/META-INF/beans.xml]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/org/drools/workbench/screens/appformer/extensions/AppformerExtensionsEditorClient.gwt.xml [hunk 28]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!--
+  ~ Copyright 2012 Red Hat, Inc. and/or its affiliates.
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/main/resources/org/drools/workbench/screens/appformer/extensions/AppformerExtensionsEditorClient.gwt.xml]: 2017

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/test/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenterTest.java [hunk 29]
```diff
@@ -0,0 +1,119 @@
+package org.drools.workbench.screens.appformer.extensions.client.editor;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/test/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenterTest.java]: License?

File: drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/test/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenterTest.java [hunk 30]
```diff
@@ -0,0 +1,119 @@
...
+        perspectiveExtensionServiceMock = new CallerMock<>(perspectiveExtensionService);
+        layoutServicesMock = new CallerMock<>(layoutServices);
+        presenter = spy(new PerspectiveExtensionEditorPresenter(view,
+                                                                perspectiveExtensionResourceType,
+                                                                perspectiveExtensionServiceMock,
+                                                                layoutServicesMock,
+                                                                layoutEditorPresenter) {
+            @Override
+            LayoutDragComponentGroup lookupPerspectiveDragComponents() {
+                return mock(
```
Comment (manstis) [inline drools-wb-screens/drools-wb-appformer-extensions-editor/drools-wb-appformer-extensions-editor-client/src/test/java/org/drools/workbench/screens/appformer/extensions/client/editor/PerspectiveExtensionEditorPresenterTest.java]: Formatting

**Pull request closed without merging**